README.md
# Ruby Sunlight client

## Description

A Ruby wrapper for the Sunlight Congress API. From the [full API documentation](http://sunlightlabs.github.com/congress/):

> A live JSON API for the people and work of Congress, provided by the Sunlight Foundation.

Find who represents you:

* Look up legislators by location or by zip code.
* Official Twitter, YouTube, and Facebook accounts.
* Committees and subcommittees in Congress, including memberships and rankings.

Or Congress' daily work:

* All introduced bills in the House and Senate, and what occurs to them (updated daily).
* Full text search over bills, with powerful Lucene-based query syntax.
* Real time notice of votes, floor activity, and committee hearings, and when bills are scheduled for debate.

## Installation

```
gem install sunlight
```

The sunlight gem works in Ruby 1.8 or 1.9.

## Set Up

First, register for an API key](http://services.sunlightlabs.com/api/register/] 

Then, you'll want to stick the following lines somewhere in your Ruby environment:

```ruby
require 'sunlight'
Sunlight::Base.api_key = 'your-api-key'
```

## Usage

The Sunlight gem fully wraps the Sunlight Labs API.

### Legislator

Time to get to the good stuff. The most useful method is @Legislator#all_for@:

```ruby
congresspeople = Sunlight::Legislator.all_for(:address => "123 Fifth Ave New York, NY 10003")
senior_senator = congresspeople[:senior_senator]
junior_senator = congresspeople[:junior_senator]
representative = congresspeople[:representative]

junior_senator.firstname          # returns "Kirsten" 
junior_senator.lastname           # returns "Gillibrand"   
junior_senator.congress_office    # returns "531 Dirksen Senate Office Building"
junior_senator.phone              # returns "202-224-4451"  
```

Note that you should make the best attempt to get a full street address, as that is geocoded behind the scenes into a lat/long pair. If all you have is a five-digit zip code, you should not use @Legislator#all_for@, instead opting for @Legislator#all_in_zipcode@ (see below). If you pass in a zip+4, then go ahead and use @Legislator#all_for@.

So @Legislator#all_for@ returns a hash of @Legislator@ objects, and the keys are @:senior_senator@, @:junior_senator@, and @:representative@. Make sure to review all the available fields from the "Sunlight Labs API":http://services.sunlightlabs.com/api/docs/legislators/. You can also pass in a lat/long pair:

```ruby
congresspeople = Sunlight::Legislator.all_for(:latitude => 33.876145, :longitude => -84.453789)
```

This bypasses the geocoding necessary by the Google Maps API. For social networks and other applications with a User object, it makes sense to geocode the user's address up front and save the lat/long data in the local database. Then, use the lat/long pair instead of address, which cuts a substantial bit of time from the @Legislator#all_for@ request since the Google Maps API Geocoding function doesn't have to be called.

Have a five-digit zip code only? You can use the @Legislator#all_in_zipcode@ method, but keep in mind that a single zip may have multiple U.S. Representatives, as congressional district lines frequently divide populous zip codes. Unlike @Legislator#all_for@, this method returns an array of legislators, and it'll be up to you to parse through them (there will be a senior senator, a junior senator, and one or more representatives).

```ruby
members_of_congress = Sunlight::Legislator.all_in_zipcode(90210)

members_of_congress.each do |member|
	# do stuff
end
```

You can also use the @Legislator#all_where@ method for searching based on available fields. Again, you'll get back an array of @Legislator@ objects:

```ruby
johns = Sunlight::Legislator.all_where(:firstname => "John")
floridians = Sunlight::Legislator.all_where(:state => "FL")
dudes = Sunlight::Legislator.all_where(:gender => "M")

johns.each do |john|
  # do stuff
end
```

Lastly, to provide your users with a name search functionality, use @Legislator#search_by_name@, which uses fuzzy matching to compensate for nicknames and misspellings. So "Ed Markey" (real name Edward Markey) and "Jack Murtha" (real name John Murtha) will return the correct matches. You can specify a higher confidence threshold (default set to 0.80) if you feel that the matches being returned aren't accurate enough. This also returns an array of @Legislator@ objects:

```ruby
legislators = Sunlight::Legislator.search_by_name("Ed Markey")
legislators = Sunlight::Legislator.search_by_name("Johnny Boy Kerry", 0.91)
```


### District

There's also the @District@ object. @District#get@ takes in either lat/long or an address and does it's best to return the correct Congressional District:

```ruby
district = Sunlight::District.get(:latitude => 33.876145, :longitude => -84.453789)
district.state            # returns "GA"
district.number           # returns "6"

district = Sunlight::District.get(:address => "123 Fifth Ave New York, NY") 
```

Finally, two more methods, @District.all_from_zipcode@ and @District.zipcodes_in@, help you out when you want to get all districts in a given zip code, or if you want to get back all zip codes in a given district.

```ruby
districts = Sunlight::District.all_from_zipcode(90210)    # returns array of District objects
zipcodes = Sunlight::District.zipcodes_in("NY", "10")     # returns array of zip codes as strings  ["11201", "11202", "11203",...]
```


### Committees

Members of Congress sit on all-important @Committees@, the smaller bodies that hold hearings and are first to review legislation.

The @Committee@ object has three identifying fields, and an array of subcommittees, which are @Committee@ objects themselves. To get all the committees for a given chamber of Congress:

```ruby
committees = Sunlight::Committee.all_for_chamber("Senate") # or "House" or "Joint"
some_committee = committees.last
some_committee.name       # "Senate Committee on Agriculture, Nutrition, and Forestry"
some_committee.id         # "SSAF"
some_committee.chamber    # "Senate"

some_committee.subcommittees.each do |subcommittee|
  # do some stuff...
end
```

The @Committee@ object also keeps a collection of members in that committee, but since that's an API-heavy call, it must be done for each Committee one at a time:

```ruby
committees = Sunlight::Committee.all_for_chamber("Senate") # or "House" or "Joint"
some_committee = committees.last    # some_committee.members starts out as nil
some_committee.load_members         # some_committee.members is now populated
some_committee.members.each do |legislator|
  # do some stuff...
end
```

Coming from the opposite direction, the @Legislator@ object has a method for getting all committees for that particular Legislator, returning an array of @Committee@ objects:

```ruby
legislators = Sunlight::Legislator.search_by_name("Byron Dorgan")
legislator = legislators.first
legislator.committees.each do |committee|
  # do some stuff...
end
```


## License

See the terms of usage for the [Sunlight Foundation's APIs](http://services.sunlightlabs.com/accounts/register/#tos).

Copyright &copy; 2009 by Luigi Montanez and the Sunlight Foundation. See LICENSE for more information.


CHANGES.md
### 2.0.0 / In-progress

* Moved old classes and methods to `Sunlight::Deprecated`:
  * `Sunlight::Legislator` -> `Sunlight::Deprecated::Legislator`
  * `Sunlight::District` -> `Sunlight::Deprecated::District`
  * `Sunlight::Committee` -> `Sunlight::Deprecated::Committee`
* Remove geocoding functions, no longer using Google Maps geocoding API (which is prohibited for applications not embedding a Google Map)
* Remove a few fields where even the old Congress API has dropped support or stopped updating the info: 
  * `votesmart_id`
  * `congresspedia_url`
  * `email`

### 1.1.0 / 2010-05-18

* Added Gemfile (thanks Bobby Wilson)
* Removed support for deprecated API methods (thanks Steve Gass)
* Added new fields to match API fields (thanks Bobby Wilson)
* Cleaned up Rakefile (thanks Chris A)

### 1.0.7 / 2010-06-11

* Support for Ruby 1.9.x, specifically the JSON gem (thanks to Brian Cardarella)

### 1.0.6 / 2010-05-25

* Added committee.rb to the gemspec (thanks to Dan Melton)

### 1.0.5 / 2009-10-27

* Implement birthdate on Legislator as a Time object

### 1.0.4 / 2009-10-02

* Implement youtube_id for YouTube usernames

### 1.0.3 / 2009-09-03

* Be a good Ruby citizen and remove require 'rubygems'

### 1.0.2 / 2009-09-03

* Ensure that Base gets loaded first.

### 1.0.1 / 2009-07-28

* Account for the API returning district numbers as "01" instead of "1"

### 1.0 / 2009-06-25

* Move to the sunlightlabs GitHub account
* Fully support the Sunlight Labs API, notably the Committee methods

### 0.9.0 / 2009-03-15

* Warning: This release is not backwards-compatible!
* Change loading behavior of base functionality, works better with Rails and Merb
* Sunlight::SunlightObject is now Sunlight::Base
* For set up, Sunlight.api_key= is now Sunlight::Base.api_key=
* For set up, using "include 'Sunlight'" is no longer recommended
* Correct usage is Sunlight::Legislator.all_for(...) instead of just Legislator.all_for(...)
* Credit to Rue the Ghetto (rughetto on GitHub) and Eric Mill for inspiring the improvements above
* Add support for senate_class ("I", "II", or "III") and in_office (0 or 1) on Legislator
* Add support for Lobbyists, Filings, and Issues
* Huge credit to mindleak on GitHub for Lobbyist-related functionality

### 0.2.0 / 2009-03-01

* Add support for twitter_id and youtube_url on Legislator
* Add Legislator#search_by_name for fuzzy name searching
* Add Legislator#all_in_zipcode for better lookups on a five-digit zip code
* Raise exception when API Key isn't set
* Credit for various fixes goes to GitHub users pengwynn, hoverbird, and wilson

### 0.1.0 / 2008-08-20

* Initial version


lib/sunlight/deprecated/legislator.rb
require 'time'

module Sunlight
  module Deprecated

    class Legislator < Base


      attr_accessor :title, :firstname, :middlename, :lastname, :name_suffix, :nickname,
                    :party, :state, :district, :in_office, :gender, :phone, :fax, :website, :webform,
                    :congress_office, :bioguide_id, :fec_id,
                    :govtrack_id, :crp_id, :twitter_id, :youtube_url, :facebook_id,
                    :senate_class, :birthdate, :fuzzy_score

      # Takes in a hash where the keys are strings (the format passed in by the JSON parser)
      #
      def initialize(params)
        params.each do |key, value|
          value = Time.parse(value) if key == "birthdate" && value && value.size > 0
          instance_variable_set("@#{key}", value) if Legislator.instance_methods.map { |m| m.to_sym }.include? key.to_sym
        end
      end
      
      # Convenience method for getting out the youtube_id from the youtube_url
      def youtube_id
        /http:\/\/(?:www\.)?youtube\.com\/(?:user\/)?(.*?)\/?$/.match(youtube_url)[1] unless youtube_url.nil?
      end
      
      # Get the committees the Legislator sits on
      #
      # Returns:
      #
      # An array of Committee objects, each possibly
      # having its own subarray of subcommittees
      def committees
        url = Sunlight::Base.construct_url("committees.allForLegislator", {:bioguide_id => self.bioguide_id})

         if (result = Sunlight::Base.get_json_data(url))
           committees = []
           result["response"]["committees"].each do |committee|
             committees << Sunlight::Deprecated::Committee.new(committee["committee"])
           end
         else
           nil # appropriate params not found
         end
         committees
      end


      #
      # Useful for getting the exact Legislators for a given district.
      #
      # Returns:
      #
      # A Hash of the three Members of Congress for a given District: Two
      # Senators and one Representative.
      #
      # You can pass in lat/long or address. The district will be
      # determined for you:
      #
      #   officials = Legislator.all_for(:latitude => 33.876145, :longitude => -84.453789)
      #   senior = officials[:senior_senator]
      #   junior = officials[:junior_senator]
      #   rep = officials[:representative]
      #
      def self.all_for(params)

        if (params[:latitude] and params[:longitude])
          Legislator.all_in_district(District.get(:latitude => params[:latitude], :longitude => params[:longitude]))
        else
          nil # appropriate params not found
        end

      end


      #
      # A helper method for all_for. Use that instead, unless you 
      # already have the district object, then use this.
      #
      # Usage:
      #
      #   officials = Sunlight::Deprecated::Legislator.all_in_district(District.new("NJ", "7"))
      #
      def self.all_in_district(district)

        senior_senator = Legislator.all_where(:state => district.state, :district => "Senior Seat").first
        junior_senator = Legislator.all_where(:state => district.state, :district => "Junior Seat").first
        representative = Legislator.all_where(:state => district.state, :district => district.number).first

        {:senior_senator => senior_senator, :junior_senator => junior_senator, :representative => representative}

      end


      #
      # A more general, open-ended search on Legislators than #all_for.
      # See the Sunlight API for list of conditions and values:
      #
      # http://services.sunlightlabs.com/api/docs/legislators/
      #
      # Returns:
      #
      # An array of Legislator objects that matches the conditions
      #
      # Usage:
      #
      #   johns = Sunlight::Deprecated::Legislator.all_where(:firstname => "John")
      #   floridians = Sunlight::Deprecated::Legislator.all_where(:state => "FL")
      #   dudes = Sunlight::Deprecated::Legislator.all_where(:gender => "M")
      #
      def self.all_where(params)

        url = construct_url("legislators.getList", params)

        if (result = get_json_data(url))

          legislators = []
          result["response"]["legislators"].each do |legislator|
            legislators << Legislator.new(legislator["legislator"])
          end

          legislators

        else  
          nil
        end # if response.class

      end
      
      #
      # When you only have a zipcode, use this.
      # It specifically accounts for the case where more than one Representative's district
      # is in a zip code.
      # 
      #
      # Returns:
      #
      # An array of Legislator objects
      #
      # Usage:
      #
      #   legislators = Sunlight::Deprecated::Legislator.all_in_zipcode(90210)
      #
      def self.all_in_zipcode(zipcode)

        url = construct_url("legislators.allForZip", {:zip => zipcode})
        
        if (result = get_json_data(url))

          legislators = []
          result["response"]["legislators"].each do |legislator|
            legislators << Legislator.new(legislator["legislator"])
          end

          legislators

        else  
          nil
        end # if response.class

      end # def self.all_in_zipcode
      
      
      # 
      # Fuzzy name searching. Returns possible matching Legislators 
      # along with a confidence score. Confidence scores below 0.8
      # mean the Legislator should not be used.
      #
      # The API documentation explains it best:
      # 
      # http://services.sunlightlabs.com/docs/congressapi/legislators.search/
      #
      # Returns:
      #
      # An array of Legislators, with the fuzzy_score set as an attribute
      #
      # Usage:
      #
      #   legislators = Sunlight::Deprecated::Legislator.search_by_name("Teddy Kennedey")
      #   legislators = Sunlight::Deprecated::Legislator.search_by_name("Johnny Kerry", 0.9)
      #
      def self.search_by_name(name, threshold='0.8')
        
        url = construct_url("legislators.search", {:name => name, :threshold => threshold})
        
        if (response = get_json_data(url))
          
          legislators = []
          response["response"]["results"].each do |result|
            if result
              legislator = Legislator.new(result["result"]["legislator"])
              fuzzy_score = result["result"]["score"]
              
              if threshold.to_f < fuzzy_score.to_f
                legislator.fuzzy_score = fuzzy_score.to_f
                legislators << legislator
              end
            end
          end
          
          if legislators.empty?
            nil
          else
            legislators 
          end
          
        else
          nil
        end
        
      end 
      
    end
  end
end